File: 프로그래머스/2/12902. 3 x n 타일링/3 x n 타일링.py
def solution(n):
    mod = 1000000007
    
    if n % 2 == 1:
        return 0
    
    d = [0] * (n + 1)
    d[2] = 3
    d[4] = 11
    
    for i in range(6, n + 1, 2):
        d[i] = (3 * d[i-2] + 2 * sum(d[2:i-2]) + 2) % mod
    
    return d[n]

File: 프로그래머스/2/12902. 3 x n 타일링/3 x n 타일링.java
class Solution {

    private static final long MOD = 1_000_000_007L;

    public int solution(int n) {
        if (n % 2 == 1) {
            return 0;
        }

        if (n == 0) {
            return 1;
        }

        if (n == 2) {
            return 3;
        }

        long previousPrevious = 1;
        long previous = 3;

        for (int i = 4; i <= n; i += 2) {
            long current = (4 * previous - previousPrevious + MOD) % MOD;

            previousPrevious = previous;
            previous = current;
        }

        return (int) previous;
    }
}